docs/architecture/adr-073-libp2p.md

```diff
@@ -0,0 +1,189 @@
+constructs (reactors), the routing layer (`Router`) and the lower
+level connection and peer management code make this operation
+relatively straight forward from an implementation perspective. The
+goal in this design is to minimize the impact to the reactors
+(potentially entirely,) and completely remove the lower level
+aspects (e.g. `Transport`, `Connection` and `PeerManager`) using the
+separation afforded by the `*Router` layer. The current state of the
+code makes these changes relatively surgical, and limited to a small
+number of methods:
+
+- `p2p.Router.OpenChannel` will still return a `Channel` structure
+  which will continue to serve as a pipe between the reactors and the
+  `Router`. The implementation will no longer need the queue
+  implementation and this operation will instead start goroutines that
+  are responsible for routing the messages from the channel to libp2p
+  fundamentals, replacing the current `p2p.Router.routeChannel`.
+
+- The current `p2p.Rotuer.dialPeers` and `p2p.Router.acceptPeers`,
+  which are responsible for establishing outbound and inbound
+  connections respectively, will be removed, along with
+  `p2p.Router.openConnection` and the libp2p connection manager will
+  be responsible for maintaining network connectivity.
+
+- The `p2p.Channel` interface should change to replace the golang
+  channels with a more functional interface for sending messages. The
+  new methods on this object, will take contexts to support safe
+  cancellation and return errors, and generally block rather than
+  return asynchronously. The `Out` channel through which
+  reactors send messages to Peers, will be replaced by a `Send`
+  method, and the Error channel will be replaced by an `Error`
+  method. 
+
+- Rectors will have access to an interface that will allow them to
+  access information from the current state of the Peer information
+  from libp2p. This will supplant the  `p2p.PeerUpdates`
+  subscription.
+
+### Upgrade and Compatibility
+
+Because the routers and all current P2P code is in the `internal`
+package and not part of the public API, the only changes to the public
+API surface area of the tendermint will be different configuration
+file options, replacing the current P2P options with options relevant
+to libp2p.
+
+However, it will not be possible to run a network with both networking
+stacks active at once, so the upgrade to the version of Tendermint
+will need to be coordinated between all nodes of the network.
+
+## Open Questions
+
+- What is the role of Protocol Labs in the implementation of libp2p in
+  tendermint, both during the initial implementation and on an ongoing
+  basis thereafter?
+
+- Should all P2P traffic for a given node be pushed to a single topic,
+  and we assume that a topic maps to a specific ChainID, or should
+  each reactor (or type of message) have its own topic?
+
+- Tendermint presently has a concept of message-type priority, which
```

Comment:
QoS is needed for multiplexing a single tcp connection correct? If we use something like quic then another channel/stream wouldn't be able to starve another. This is how I understand quic.

Reply:
In a lot of ways the chief failing of the current system is that it pushes about 2/3rds of the QoS responsibility into the networking layer, which makes it really hard to do anything other than buffer a lot and hope. 

Since we're expecting to start off with TCP and iterate from there, I don't know that we can trust QUIC to solve this problem entirely. No matter how you abstract the socket (or groups of socket) the bits have to go over the same wire.